File: ChatJava/app/src/main/java/com/example/chatjava/viewmodel/SohbetListesiViewModel.java
package com.example.chatjava.viewmodel;



import android.os.Handler;
import android.os.Looper;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.example.chatjava.model.KonusulanKisi;
import com.example.chatjava.service.ApiService;
import com.example.chatjava.service.RetrofitClient;
import com.example.chatjava.service.response.KonusulanKisiListResponse;

import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;


public class SohbetListesiViewModel extends ViewModel {


    private final MutableLiveData<List<KonusulanKisi>> _konusulanKisiler = new MutableLiveData<>();
    public LiveData<List<KonusulanKisi>> getKonusulanKisiler() { return _konusulanKisiler; }

    private final MutableLiveData<String> _hataMesaji = new MutableLiveData<>();
    public LiveData<String> getHataMesaji() { return _hataMesaji; }


    private final ApiService api = RetrofitClient.getApiService();


    private final Handler handler = new Handler(Looper.getMainLooper());
    private Runnable yenilemeRunnable;

    public void sohbetListesiniBaslat(int kullaniciId) {
        // Önceki döngü varsa iptal et
        if (yenilemeRunnable != null) {
            handler.removeCallbacks(yenilemeRunnable);
        }

        yenilemeRunnable = new Runnable() {
            @Override
            public void run() {
                sohbetListesiniGetir(kullaniciId);
                handler.postDelayed(this, 15_000);
            }
        };
        handler.post(yenilemeRunnable);
    }


    private void sohbetListesiniGetir(int kullaniciId) {
        api.konusulanKisiler(kullaniciId)
                .enqueue(new Callback<KonusulanKisiListResponse>() {
                    @Override
                    public void onResponse(Call<KonusulanKisiListResponse> call,
                                           Response<KonusulanKisiListResponse> response) {

                        if (response.isSuccessful()
                                && response.body() != null
                                && response.body().isSuccess()) {

                            _konusulanKisiler.postValue(response.body().getKisiler());
                            _hataMesaji.postValue(null);

                        } else {
                            _hataMesaji.postValue("Liste alınamadı");
                        }
                    }

                    @Override
                    public void onFailure(Call<KonusulanKisiListResponse> call, Throwable t) {
                        _hataMesaji.postValue("Sunucu hatası: " + t.getMessage());
                    }
                });
    }

    @Override
    protected void onCleared() {
        super.onCleared();
        if (yenilemeRunnable != null) {
            handler.removeCallbacks(yenilemeRunnable);
            yenilemeRunnable = null;
        }
    }
}

File: ChatKotlin/app/src/main/java/com/example/chatkotlin/viewmodel/SohbetListesiViewModel.kt
package com.example.chatkotlin.viewmodel

import androidx.lifecycle.*
import com.example.chatkotlin.model.KonusulanKisi
import com.example.chatkotlin.service.RetrofitClient
import com.example.chatkotlin.service.response.KonusulanKisiListResponse
import kotlinx.coroutines.*
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class SohbetListesiViewModel : ViewModel() {

    private val _konusulanKisiler = MutableLiveData<List<KonusulanKisi>>()
    val konusulanKisiler: LiveData<List<KonusulanKisi>> get() = _konusulanKisiler

    private val _hataMesaji = MutableLiveData<String?>()
    val hataMesaji: LiveData<String?> get() = _hataMesaji

    private var yenilemeJob: Job? = null

    fun sohbetListesiniBaslat(kullaniciId: Int) {
        yenilemeJob?.cancel()

        yenilemeJob = viewModelScope.launch {
            while (isActive) {
                sohbetListesiniGetir(kullaniciId)
                delay(15_000)
            }
        }
    }

    private fun sohbetListesiniGetir(kullaniciId: Int) {
        RetrofitClient.apiService.konusulanKisiler(kullaniciId)
            .enqueue(object : Callback<KonusulanKisiListResponse> {
                override fun onResponse(
                    call: Call<KonusulanKisiListResponse>,
                    response: Response<KonusulanKisiListResponse>
                ) {
                    if (response.isSuccessful && response.body()?.success == true) {
                        _konusulanKisiler.postValue(response.body()?.kisiler ?: emptyList())
                        _hataMesaji.postValue(null)
                    } else {
                        _hataMesaji.postValue("Liste alınamadı")
                    }
                }

                override fun onFailure(call: Call<KonusulanKisiListResponse>, t: Throwable) {
                    _hataMesaji.postValue("Sunucu hatası: ${t.message}")
                }
            })
    }

    override fun onCleared() {
        super.onCleared()
        yenilemeJob?.cancel()
    }
}
